src/theory/quantifiers/trigger.h

```diff
@@ -236,75 +260,98 @@ class Trigger {
-  static void getTriggerVariables( Node icn, Node f, std::vector< Node >& t_vars );
+  /** get all variables that E-matching can instantiate from a subterm n.
+  * This returns the union of all free variables in usable triggers that are subterms of n.
+  */
+  static void getTriggerVariables( Node n, Node f, std::vector< Node >& t_vars );
 protected:
-  /** trigger constructor */
+  /** trigger constructor, intentionally protected (use Trigger::mkTrigger). */
   Trigger( QuantifiersEngine* ie, Node f, std::vector< Node >& nodes );
-
-  /** is subterm of trigger usable */
+  /** is subterm of trigger usable (helper function for isUsableTrigger) */
   static bool isUsable( Node n, Node q );
+  /** returns an equality that is equivalent to the equality eq and 
+  * is a usable trigger for q if one exists, otherwise returns Node::null().
+  */
   static Node getIsUsableEq( Node q, Node eq );
+  /** returns whether n1 == n2 is a usable (relational) trigger for q. */
   static bool isUsableEqTerms( Node q, Node n1, Node n2 );
-  /** collect all APPLY_UF pattern terms for f in n */
+  /** helper function for collectPatTerms */
   static void collectPatTerms2( Node q, Node n, std::map< Node, std::vector< Node > >& visited, std::map< Node, TriggerTermInfo >& tinfo, 
                                 quantifiers::TriggerSelMode tstrt, std::vector< Node >& exclude, std::vector< Node >& added,
                                 bool pol, bool hasPol, bool epol, bool hasEPol, bool knowIsUsable = false );
                                 
   /** add an instantiation (called by InstMatchGenerator) 
   * This calls d_quantEngine->addInstantiation(...) for instantiations associated with m.
   * Typically, m is associated with a single instantiation, but in some cases (e.g. higher-order)
-  * we may modify m before sending it.
+  * we may modify m before calling d_quantEngine->addInstantiation(...).
   */
   virtual bool sendInstantiation( InstMatch& m );
-  /** the nodes comprising this trigger */
+  /** The nodes comprising this trigger. */
   std::vector< Node > d_nodes;
-  /** the quantifiers engine */
+  /** The quantifiers engine associated with this trigger. */
   QuantifiersEngine* d_quantEngine;
-  /** the quantifier this trigger is for */
+  /** The quantifier this trigger is for. */
   Node d_f;
   /** match generator
-  * This is the class that implements the underlying matching algorithm
+  * This is backend utility that implements the underlying matching algorithm
   * associated with this trigger.
   */
   IMGenerator* d_mg;
 }; /* class Trigger */
 
-/** a trie of triggers 
-* This stores an index of triggers, 
+/** A trie of triggers.
+* 
+* This class is used to cache all Trigger objects that are generated in the 
+* current context. We index Triggers in this data structure based on the 
+* value of Trigger::d_nodes. When a Trigger is added to this data structure,
+* this Trie assumes responsibility for deleting it.
 */
```

Comment:
Does it make sense (in a separate issue) to consider having a generic trie template?  It seems to be a commonly used data structure.

Reply:
I prefer having multiple copies of this so that custom optimizations can be made for the particular instance